test/hotspot/jtreg/serviceability/jvmti/vthread/FollowReferences/VThreadStackRefTest.java

```diff
@@ -0,0 +1,193 @@
+ * @run main/othervm/native
+ *      -Djava.util.concurrent.ForkJoinPool.common.parallelism=1
+ *      -agentlib:VThreadStackRefTest
+ *      VThreadStackRefTest
+ */
+
+import java.lang.ref.Reference;
+import java.util.stream.Stream;
+import java.util.concurrent.CountDownLatch;
+
+/*
+ * The test verifies JVMTI FollowReferences function reports references from
+ * mounted and unmounted virtual threads and reports correct thread id
+ * (for mounted vthread it should be vthread id, and not carrier thread id).
+ * Additionally tests that references from platform threads aree reported correctly
+ * and that references from terminated vthread are not reported.
+ * To get both mounted and unmounted vthreads the test:
+ * - limits the number of carrier threads to 1;
+ * - starts vthread that creates a stack local and JNI local
+ *   and then waits in CountDownLatch.await();
+ * - starts another vthread that create stack local and JNI local (on top frame)
+ *   and waits in native to avoid unmounting.
+ */
+public class VThreadStackRefTest {
+
+    // The flag is set by createObjAndWait method.
+    static volatile boolean mountedVthreadReady = false;
+
+    public static void main(String[] args) throws InterruptedException {
+        CountDownLatch dumpedLatch = new CountDownLatch(1);
+
+        CountDownLatch unmountedThreadReady = new CountDownLatch(1);
+        // Unmounted virtual thread with stack local and JNI local.
+        Thread vthreadUnmounted = Thread.ofVirtual().start(() -> {
+            Object referenced = new VThreadUnmountedReferenced();
+            System.out.println("created " + referenced.getClass());
+            createObjAndCallback(VThreadUnmountedJNIReferenced.class,
+                new Runnable() {
+                    public void run() {
+                        unmountedThreadReady.countDown();
+                        await(dumpedLatch);
+                    }
+                });
+            Reference.reachabilityFence(referenced);
+        });
+        // Wait until unmounted thread is ready.
+        unmountedThreadReady.await();
+
+        // Ended virtual thread with stack local - should not be reported.
+        Thread vthreadEnded = Thread.ofVirtual().start(() -> {
+            Object referenced = new VThreadUnmountedEnded();
+            System.out.println("created " + referenced.getClass());
+            Reference.reachabilityFence(referenced);
+        });
+        // Make sure this vthread has exited so we can test
+        // that it no longer holds any stack references.
+        vthreadEnded.join();
+
+        // Mounted virtual thread with stack local and JNI local on top frame.
+        Thread vthreadMounted = Thread.ofVirtual().start(() -> {
+            Object referenced = new VThreadMountedReferenced();
+            System.out.println("created " + referenced.getClass());
+            createObjAndWait(VThreadMountedJNIReferenced.class);
+            Reference.reachabilityFence(referenced);
+        });
```

Comment:
This code used to use a java loop to keep busy, but now it relies on a sleep loop in native code. Was the java loop problematic?

Reply:
No. I had a failure of the test due racing, but I believe the reason was lack of synchronization.
I decided to simplify the test - 2 virtual threads are enough and as the test verifies "JNI local on top frame" case, it needs block in native call anyway, so I use it to prevent unmount too.